src/bootstrap/src/core/builder.rs

```diff
@@ -438,8 +438,28 @@ impl StepDescription {
             }
         }
 
-        // strip CurDir prefix if present
-        let mut paths: Vec<_> = paths.iter().map(|p| p.strip_prefix(".").unwrap_or(p)).collect();
+        // Attempt to resolve paths to be relative to the builder source directory.
+        let paths: Vec<_> = paths
+            .iter()
+            .map(|p| {
+                // If the path does not exist, it may represent the name of a Step, such as `tidy` in `x test tidy`
+                if !p.exists() {
+                    return p.clone();
+                }
+
+                // Make the path absolute, strip the prefix, and convert to a PathBuf.
+                match std::path::absolute(p) {
+                    Ok(p) => p.strip_prefix(&builder.src).unwrap_or(&p).to_path_buf(),
+                    Err(e) => {
+                        eprintln!("ERROR: {:?}", e);
+                        panic!("Due to the above error, failed to resolve path: {:?}", p);
+                    }
+                }
+            })
+            .collect();
+
+        // Convert Vec<PathBuf> to `Vec<&Path>`
+        let mut paths: Vec<&Path> = paths.iter().map(|p| p.as_ref()).collect();
```

Comment:
You can convert `Vec<&Path>` argument types to `Vec<PathBuf>`, so we don't need this extra iteration.

Reply:
Makes sense to me. I just push a change that removes this line and updates the argument types for `remap_paths()`, `PathSet::intersection_removing_matches()`, and `ShouldRun::pathset_for_paths_removing_matches()`.